src/main/scala/viper/gobra/frontend/Config.scala

```diff
@@ -342,98 +366,105 @@ class ScallopGobraConfig(arguments: Seq[String], isInputOptional: Boolean = fals
-      }
-
-      // perform the following checks:
-      // - validate fileOpt using includeOpt
-      // - convert fileOpt using includeOpt
-      //  - result should be non-empty, exist, be files and be readable
-      val input: List[String] = inputOpt.getOrElse(List())
-      for {
-        convertedFiles <- checkConversion(input)
-        _ <- atLeastOnePath(convertedFiles)
-        _ <- pathsExist(convertedFiles)
-        _ <- pathsAreFilesOrDirectories(convertedFiles)
-        _ <- pathsAreReadable(convertedFiles)
-      } yield ()
-  }
-
+  validateFilesExist(cutInput)
+  validateFilesIsFile(cutInput)
+  validateFileExists(directory)
+  validateFileIsDirectory(directory)
   if (!skipIncludeDirChecks) {
     validateFilesExist(include)
     validateFilesIsDirectory(include)
   }
+  InputConverter.validate(cutInput, directory, recursive, includePackages, excludePackages)
 
-  // List of input arguments together with their specified line numbers.
-  // Specified line numbers are removed from their corresponding input argument.
-  val cutInputWithIdxs: ScallopOption[List[(String, List[Int])]] = input.map(_.map{ arg =>
-    val pattern = """(.*)@(\d+(?:,\d+)*)""".r
-    arg match {
-      case pattern(prefix, idxs) =>
-        (prefix, idxs.split(',').toList.map(_.toInt))
+  private object InputConverter {
+    /**
+      * stores the computed inputPackageMap after validating them to avoid a recomputation. `None` indicates that either validation
+      * has not yet happened or that validation has failed
+      */
+    var inputPackageMapOpt: Option[Map[PackageInfo, Vector[Source]]] = None
+    /** can only be accessed after calling `validate` */
+    lazy val inputPackageMap: Map[PackageInfo, Vector[Source]] = inputPackageMapOpt.get
 
-      case _ => (arg, List.empty[Int])
+    /**
+      * we distinguish between two modes:
+      * - input mode in which the user simply provides all files that should be verified
+      * - directory mode in which Gobra searches for the files to be verified
+      * if inputs are provided, it checks that there is at least one source after resolving the sources.
+      * if a directory is provided, it checks that at least 1 package remains to be verified after applying the desired filters */
+    def validate(cutInput: ScallopOption[List[File]], directory: ScallopOption[File], recursive: ScallopOption[Boolean], includePackages: ScallopOption[List[String]], excludePackages: ScallopOption[List[String]]): Unit = {
+      validateOpt (cutInput, directory, recursive, includePackages, excludePackages) ((cutInputOpt, dirOpt, recOpt, includeOpt, excludeOpt) => {
+        // `requireOne(input, directory)` above makes sure that only `input` or `directory` is provided but not both.
+        // Thus, we can simply take their concatenation
+        // Note however that both options might not exist if the flag `isInputOptional` is set to true
```

Comment:
These two flags are only used in the context of parsing in-file configs because you would otherwise be forced to always specify at least one input (to make the input check happy). In the case of include dirs, we agreed that they should be relative to the current file's position, which interferes with the way we normally check the validity of include dirs (i.e. we check whether the specified directory exists, which is relative to the current directory though).
My first attempt was also to skip the entire validation when the flags are set to true, but reading `inputPackageMap` then fails because there I explicitly check that validation has happened